accepted/2022/simplify-rid-model.md

```diff
@@ -0,0 +1,252 @@
+
+| RID        | Description of scope |
+|------------|----------------------|
+| any        | The root RID, which is compatible with all other RIDs. |
+| alpine     | All Alpine versions and architecture builds. |
+| alpine-x64 | All Alpine versions, for x64. Other processor-types are also supported. |
+| alpine3.16-arm64 | Alpine 3.16, for Arm64. Other processor-types are also supported.|
+| centos     | CentOS has a similar scheme as Alpine. |
+| debian     | Debian has a similar scheme as Alpine. |
+| osx        | macOS has a similar scheme as Alpine. |
+| rhel       | Red Hat Enterprise Linux has a similar scheme as Alpine. |
+| tizen      | Tizen has a similar scheme as Alpine. |
+| ubuntu     | Ubuntu has a similar scheme as Alpine. |
+
+Note: Many other Linux distros are represented in `runtimes.json`.
+
+Note: `runtimes.json` will be frozen, which means that RID schemes only supported by this file are abandoned. For example, RIDs that include OS versions will not be updated going forward.
+
+## Source-build
+
+A major design tenet of the RID system is maximizing portability of code. This is particularly true for Linux. That makes sense from the perspective of Microsoft wanting to make one build of .NET available across many Linux distros, separately for both glibc and musl. It makes less sense for distros themselves building .NET from source and publishing binaries to a distro package feed.
+
+We sometimes refer to `linux-x64` (and equally to `linux-arm64` and `linux-x86`) as the "portable Linux RID". As mentioned in the libc section, this RID establishes a wide glibc compatibility range. In addition, the RID also establishes broad compatibility across distros (and their associated versions) for .NET dependencies, like OpenSSL and ICU. The way this particular form of compatibility shows up is observable but is nuanced (and isn't explained here).
+
+We'll use Red Hat as an example of an organization that builds .NET from source, to complete this discussion. They have raised concerns on [source-build not playing nicely with `linux-x64`](https://github.com/dotnet/runtime/pull/62942#issuecomment-1056942235).
+
+For Red Hat, it makes sense to accept and support `linux-x64` NuGet assets, but not to produce or offer them. Instead, Red Hat would want to produce `rhel-x64` runtime pack assets. It's easiest to describe this design-point in terms of concrete scenarios.
+
+**NuGet packages** -- NuGet authors will typically value breadth and efficiency. For example, a package like `SkiaSharp` might target `linux-x64`, but not `rhel-x64`. There is no technical reason for that package to include distro-specific assets. Also, if the author produced a `rhel-x64` asset, they would need to consider producing an `ubuntu-x64` and other similar assets for other distros, and that's not scalable. The expectation is that the `rhel-x64` .NET supports `linux-x64` NuGet assets, enabling NuGet authors to target a minimal set of RIDs.
```

Comment:
I feel like this document generally, and this section specifically, is making some assumptions that I don't think are justified. There are absolutely scenarios where distro-specific assets need to be shipped in a NuGet package.

As an example of this, take a look at https://www.nuget.org/packages/LibGit2Sharp.NativeBinaries/2.0.306

LibGit2Sharp is a managed wrapper around the native libgit2 library and has to accommodate the native dependencies of libgit2. One of the dependencies is OpenSSL, which varies wildly among all the distros that .NET supports.

In order to have a reasonable chance of having a native binary that just works on all platforms, I have had to spend a lot of time understanding the RID graph and shipping enough distro-specific binaries to cover all the supported Linux distros

I will point out that libgit2 has recently changed how they are binding to OpenSSL which has let me simplify things down to more of an "ideal" situation (see the [newer package](https://www.nuget.org/packages/LibGit2Sharp.NativeBinaries/2.0.315)), but that was largely out of my control since I'm not a libgit2 maintainer.

If the ability to ship distro-specific assets in a NuGet package and have the proper one selected had not been a feature that I could count on, then I would not have been able to support LibGit2Sharp on Linux at all for the past several years. Maybe I'm misunderstanding something, but it seems like this proposal is removing this feature.

Reply:
Mind you, https://github.com/libgit2/libgit2sharp/pull/1714 changed the model used by libgit2sharp. AFAIK, the change basically makes libgit2sharp try and load all native libgit2 libraries one by one until one of them can be loaded successfully (ie, links to a version of OpenSSL available on the system). The actual name (and RID) is not too relevant in this scenario.

I feel like your use-case would still work if nuget did not "know" about distro-specific RIDs.

Reply:
This is great feedback. We discussed it.

There are three models:

- Don't support specific distros as a general concept (the current proposal).
- Fully support specific distros as a general concept (some variation on status-quo).
- Adopt a model that is a variation on the Python manylinux wheel plan, including support for OpenSSL.

I like what @omajid is proposing. It's the same approach we use for OpenSSL. We don't use RIDs. Also, the current proposal works for the vast majority of cases and is a massive simplification. I'm hoping we can keep it.

Reply:
libgit2sharp was suffering from the rid scalability issue: it was unable to deal with unknown rids.

Using `NativeLibrary.SetDllImportResolver` it now tries the different `so` files. The code is here: https://github.com/libgit2/libgit2sharp/blob/97bee65fd296f1c7dd2d1d64581c170f45b584e1/LibGit2Sharp/Core/NativeMethods.cs#L99-L121

This way, the library can have its own logic for picking `so` files. It's not limited to/by rids.

Reply:
Yes, this logic was added, but I consider it a fallback solution, and not something desirable to rely on as the primary logic, for a couple of reasons:
- It only works if the entire runtimes folder is shipped as part of the application, which is only going to be true if you use one of the cross-platform publishing options
- Having to attempt to load every binary until you find one that works seems like it would have a negative impact on startup perf.

If libraries no longer can assume that the best native asset is selected for them by the framework/NuGet, then that means they have to manually ensure that all binaries are included in the publish output to be able to select from them at runtime. That can pretty massively increase the size of the output. Using LibGit2Sharp.NativeBinaries 2.0.306 as an example, that means going from having a single ~1MB .so file in your output to having 9 different copies, and that's assuming I'd have some way to know to only copy the Linux binaries and not the entire runtimes folder.

Reply:
> It only works if the entire runtimes folder is shipped as part of the application, which is only going to be true if you use one of the cross-platform publishing options

This way of trimming it isn't scalable as you need to build and include an `so` file for every platform you want to trim.

A scalable solution is to allow to control the trimming more specifically. For example add a property to csproj that enables the user to choose the supported OpenSSL versions, and include `so` files based on that.

> Having to attempt to load every binary until you find one that works seems like it would have a negative impact on startup perf.

I doubt this has a measurable impact.

> Using LibGit2Sharp.NativeBinaries 2.0.306 as an example, that means going from having a single ~1MB .so file in your output to having 9 different copies

It seems the latest version doesn't have distro specific versions. Why did it change?

Reply:
> perf

.NET probes for OpenSSL. No one has ever submitted a perf report on that.

@bartonjs -- can you point to the probing code for OpenSSL?

Reply:
> can you point to the probing code for OpenSSL?

OpenSSL probing code in dotnet/runtime: https://github.com/dotnet/runtime/blob/main/src/native/libs/System.Security.Cryptography.Native/opensslshim.c#L54-L121

OpenSSL probing in libgit2: 
https://github.com/libgit2/libgit2/blob/82f526a9129373d899782c411774c8e396cae4e4/deps/ntlmclient/crypt_openssl.c#L63-L69 . Notice that it does not have support for probing for OpenSSL 3 yet so it is going to break once distros switch to OpenSSL 3 only. dotnet/runtime has probing for OpenSSL3 that adds a whole new level of shimming and complexity.

> perf

The probing is relatively expensive when looked at in isolation (it is file I/O). It is relatively cheap compared to e.g. all JITing that happens during startup of a typical .NET app.

> I see there is a linux-x64 asset. What's the intended behavior of that asset? By definition, it can run on any (glibc) distro.

Yes, this works like bow that since libgit2 added a configuration to probe for OpenSSL (https://github.com/libgit2/libgit2/pull/5974) and the linux-x64 binaries in the nuget package are built with this option on.

Reply:
> This way of trimming it isn't scalable as you need to build and include an so file for every platform you want to trim.

That's not how publishing works currently, though. It uses the RID graph to select the best option from the assets in the package. My goal was always to have enough coverage so that the supported .NET platforms would just work. That *was* broken sometimes due to the RID graph not being updated...

> It seems the latest version doesn't have distro specific versions. Why did it change?

I mentioned that earlier here: libgit2 has recently added a dynamic binding option for OpenSSL, which has let me simplify the package.

> I see there is a linux-x64 asset. What's the intended behavior of that asset? By definition, it can run on any (glibc) distro.

The set of binaries in the package is a result of analyzing the distros that .NET Core 3.1 supported and the RID graph to ship the fewest number of binaries with the maximum amount coverage. The binary shipped as `linux-x64` was intended to be a fallback that would work when a more specific binary was not chosen. That meant I needed to understand the glibc and OpenSSL versions in each distro to be able to construct the set of binaries I shipped.

This was a large effort, but the end result was that most/all the distros would work out of the box. However, when the RID graph was not updated for a new .NET Core release and a new distro/version incorrectly defaulted back to `linux-x64`, then that fallback would fail. That's where @tmds PR came in and added an additional fallback to try and see if one of the binaries in the package might work even though the RID graph wasn't able to select it.

Reply:
> For example add a property to csproj that enables the user to choose the supported OpenSSL versions, and include so files based on that.

The problem does not end with OpenSSL. The same problem exists for many other libraries.

I believe that the only truly scalable option for Linux ecosystem is to have building from source as an option (https://github.com/NuGet/Home/issues/9631). Building from source is capable of producing a binary that works for the specific configuration of your system without the package maintainer pre-building it for you.

Reply:
> The problem does not end with OpenSSL. The same problem exists for many other libraries.

Yes, and this was the main point of the feedback I was trying to raise here. The proposal currently assumes that there is never a reason to care about the specific distro you're running on, but that is demonstrably not true.

This all comes down to there being a disconnect between the .NET ecosystem expectations of sharing pre-compiled binaries, and the Linux ecosystem that expects you to share source and compile it on your own system. Shipping a Linux binary in a package is always going to be an uphill battle, so I would like to see the .NET side of things evolve in a way that makes it easier, not harder.

Reply:
> Notice that it does not have support for probing for OpenSSL 3 yet so it is going to break once distros switch to OpenSSL 3 only. dotnet/runtime has probing for OpenSSL3 that adds a whole new level of shimming and complexity.

Let's assume OpenSSL 3 won't be supported by a libgit2 shimming (immediately) and you want to solve it issue using distro rids.

Fedora 36 should use the OpenSSL 3 build.

Fedora 37 should be using the OpenSSL 3 build also. That doesn't work automagically today, because the rid grah doesn't describe a supports relationship between Fedora 37 and Fedora 36.

For your package to remain maintainable, this needs to be addressed (and preferably not by bloating the graph but by making the rid logic smarter).

I've only mentioned Fedora here, but OpenSSL will be adopted by Ubuntu, Debian, so you need to manage this problem for them too.

Your package now looks like this:

```
linux-x64 -> supports OpenSSL 1, 2, 3 and picks using `NativeLibrary`
fedora.36 -> OpenSSL 3
ubuntu.22.04 -> OpenSSL 3
debian.12 -> OpenSSL 3
...
```

Note that you are including the distro specific assets solely for the purpose of trimming. They bloat your package and require you to track in what version a distro adopts a new OpenSSL version.

> The proposal currently assumes that there is never a reason to care about the specific distro you're running on

`NativeLibrary` can be used, and should be used to make `linux-x64` work across a range of distros.

What is lost, is the ability to trim against a rid.

Reply:
@bording I hope this example helped illustrate how unscalable the rid mechanism is?

And, if it should stay, inheritance between successive distro versions is a must-have, which is now missing.

Reply:
> And, if it should stay, inheritance between successive distro versions is a must-have, which is now missing.

To elaborate. The rids as we have them are not scalable. The alternative to the simple model that is proposed here, is to have _semantic rids_. That means rids follow a `<distro>.<version>-<arch>` naming scheme. Information is derived from that instead of hard coding it in the json file (e.g. Fedora 36 -> Fedora 35). The json file should then only contain non-derivable information (like what distros are glibc/musl based).

This is a much more complex solution.